Add Core Project.

Stan-Kudri opened the pull request.

Stan-Kudri committed:
Solution: Core - Add; GameSnake - Rename namespace and Changed class "Speed".

1)Add new project "Core".
2)Renamed throughout the project "GameSnake" from "Components" to "ComponentsGame".
3)Changed class "Speed".

Stan-Kudri committed:
TestSnake, SpeedTest - The test worked without changing the class "Speed".

Stan-Kudri committed:
Solution: Core - moved the logic from the "GameSnake" project.

blowin commented on GameSnake/Game.cs:
```diff
@@ -6,10 +7,10 @@ namespace GameSnake
     {
-        private readonly GameMap _gameMap;
+        private readonly GameMapConsole _gameMap;
         private readonly UserInput _userInput;
```
Переделай на IUserInput, и уже ссылайся на интерфейс, так сможешь отвязаться от конкретной реализации

blowin commented on GameSnake/Game.cs:
```diff
@@ -6,10 +7,10 @@ namespace GameSnake
         private readonly UserInput _userInput;
-        private readonly Score _score;
-        private readonly Speed _speed;
+        private readonly ScoreConsole _score;
```
Наверное тут пошло что-то не так

blowin commented on GameSnake/Game.cs:
```diff
@@ -6,10 +7,10 @@ namespace GameSnake
-        private readonly Score _score;
-        private readonly Speed _speed;
+        private readonly ScoreConsole _score;
+        private readonly SpeedConsole _speed;
```
Тут тоже

Stan-Kudri committed:
Solution: Core - Remove "UserInput" and Add Interface "IUserInput"; GameSnake - Add "UserInput" inheritor of the interface "IUserInput"

Stan-Kudri committed:
Solution: Core - Add method "UpDate" from "IUserInput"; GameSnake - Changed variable types to instances from project "Core".

Stan-Kudri committed:
Solution: Core - Changed to abstract class "Speed"; GameSnake.Game - Changed the field type from "SpeedConsole" to "Speed".

blowin commented on Core/Components/GameMap.cs:
```diff
@@ -26,7 +24,7 @@ public GameMap(Border border, Snake snake)
         public bool IsGameOver() => _snake.ObstacleCollision() || _snake.Intersect();
 
-        public void ChangeSnakeDirection(UserInput direction) => _snake.Direction = direction.CurrentDirection;
+        public abstract void ChangeSnakeDirection(IUserInput direction);
```
почему нельзя в интерфейсе добавить свойство CurrentDirection?

blowin commented on Core/Game.cs:
```diff
@@ -0,0 +1,7 @@
+namespace Core
+{
+    public class Game
```
чего пустой?

blowin commented on GameSnake/Game.cs:
```diff
@@ -43,7 +43,7 @@ public void Run()
                 _gameMap.Clear();
 
-                _userInput.Update();
+                _userInput.UpDate();
```
UpDate?

Stan-Kudri committed:
Solution: Core and GameSnake - fixed problem by comments.

Stan-Kudri committed:
Solution: Core And GameSnake - Factory method implementation for "Food".

Stan-Kudri committed:
Solution, GameSnake and Core - Factory method implementation for "GameMap".

Stan-Kudri committed:
TestSnake: GameMapTest - Fixed test.

Stan-Kudri committed:
Solution: Core and GameSnake - Factory method implementation for "Speed"

Stan-Kudri committed:
Solution: Core and GameSnake - Factory method implementation for "Score"

Stan-Kudri committed:
Solution: Core and GameSnake - Factory method implementation for "IUserInput"

Stan-Kudri committed:
Solution: Game - Implementation of logic in the "Core" project.

-Solution: Core and GameSnake - Factory method implementation for "Border";
-Solution: Core and GameSnake - Factory method implementation for "Snake";
-Solution: Core and GameSnake - Add classes "GameFactory" and "GameFactoryConsole";

Stan-Kudri committed:
GameSnake: SnakeFactoryConsole - Fixed linter.

Stan-Kudri committed:
GameSnake: GameFactoryConsole - Fixed linter.

blowin commented on Core/Games/Game.cs:
```diff
@@ -0,0 +1,70 @@
+        private readonly Score _score;
+        private readonly Speed _speed;
+
+        public Game(
```
Такое количество фабрик не нужно, что-то нужно в единственном экземпляре нужно Score, Speed. Зачем тогда делать фабрику, если можно создать их выше и передать сразу готовый объект? Всё остальное тоже нужно проверить на целесообразность наличия фабрики

blowin commented on GameSnake/ComponentsGame/GameMaps/GameMapConsole.cs:
```diff
@@ -0,0 +1,41 @@
+        {
+        }
+
+        public override void Draw()
```
почему это не может быть базовом классе?

blowin commented on GameSnake/ComponentsGame/GameMaps/GameMapConsole.cs:
```diff
@@ -0,0 +1,41 @@
+            _border.Draw();
+        }
+
+        public override void Clear()
```
почему это не может быть базовом классе?

blowin commented on GameSnake/ComponentsGame/GameMaps/GameMapConsole.cs:
```diff
@@ -0,0 +1,41 @@
+{
+    public class GameMapConsole : GameMap
+    {
+        public GameMapConsole(BorderConsole border, SnakeConsole snake)
```
зачем столько перегрузок?

blowin commented on GameSnake/Games/GameFactoryConsole.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace GameSnake.Games
+{
+    public class GameFactoryConsole : GameFactory
```
мне кажется, что ты сильно переборщил с фабриками)

Stan-Kudri committed:
Solution - Removed a large number of unnecessary factories.

Stan-Kudri committed:
Solution: Deleted the factories and changed the implementation of the "GameMap" and "Game" classes.

Stan-Kudri committed:
Solution: Core and GameSnake - Removed fields from the constructor class "Game".

Stan-Kudri committed:
Solution: Core and GameSnake - Removed GameMapFactory.

blowin commented on GameSnake/Extension/SnakeFieldExtension.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public const int DividerLengthHalf = 2;
+
+        public static Snake Creator(this Border border, int length)
```
Странно, это находится не в Console проекте? Такого не должно быть

blowin commented on GameSnake/Games/GameFactoryConsole.cs:
```diff
@@ -14,25 +11,21 @@ public class GameFactoryConsole : GameFactory
         public override Game Create(
```
не уверен в целесообразности, может быть в Main

Stan-Kudri committed:
Solution - Changes on two comments.
